Sources/Algorithms/LazySplitSequence.swift

```diff
@@ -0,0 +1,321 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift Algorithms open source project
+//
+// Copyright (c) 2021 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+//
+//===----------------------------------------------------------------------===//
+
+/// A sequence that lazily splits a base sequence into subsequences separated by
+/// elements that satisfy the given `whereSeparator` predicate.
+///
+/// - Note: This type is the result of
+///
+///     x.split(maxSplits:omittingEmptySubsequences:whereSeparator)
+///     x.split(separator:maxSplits:omittingEmptySubsequences)
+///
+///   where `x` conforms to `LazySequenceProtocol`.
+public struct LazySplitSequence<Base: Sequence> {
+  internal let base: Base
+  internal let isSeparator: (Base.Element) -> Bool
+  internal let maxSplits: Int
+  internal let omittingEmptySubsequences: Bool
+}
+
+extension LazySplitSequence {
+  public struct Iterator {
+    public typealias Element = [Base.Element]
+
+    internal var base: Base.Iterator
+    internal let isSeparator: (Base.Element) -> Bool
+    internal let maxSplits: Int
+    internal let omittingEmptySubsequences: Bool
+    internal var separatorCount = 0
+    internal var sequenceLength = 0
+
+    internal init(
+      base: Base.Iterator,
+      whereSeparator: @escaping (Base.Element) -> Bool,
+      maxSplits: Int,
+      omittingEmptySubsequences: Bool
+    ) {
+      self.base = base
+      self.isSeparator = whereSeparator
+      self.maxSplits = maxSplits
+      self.omittingEmptySubsequences = omittingEmptySubsequences
+    }
+  }
+}
+
+extension LazySplitSequence.Iterator: IteratorProtocol {
+  public mutating func next() -> Element? {
+    /// Separators mark the points where we want to split (cut in two) the base
+    /// collection, removing the separator in the process.
+    ///
+    /// Each split yields two subsequences, though splitting at the start or end
+    /// of a sequence yields an empty subsequence where there were no elements
+    /// adjacent to the cut.
+    ///
+    /// Thus the maximum number of subsequences returned after iterating the
+    /// entire base collection (including empty ones, if they are not omitted)
+    /// will be at most one more than the number of splits made (equivalently,
+    /// one more than the number of separators encountered).
+    ///
+    /// The number of splits is limited by `maxSplits`, and thus may be less
+    /// than the total number of separators in the base collection.
+    ///
+    ///     [1, 2, 42, 3, 4, 42, 5].split(separator: 42,
+    ///                                   omittingEmptySubsequences: false)
+    ///     // first split -> [1, 2], [3, 4, 42, 5]
```

Comment:
Sorry, maybe I'm missing something here but what is `first split` and `last split`? For what I understood the first is a split limited by `2` and last is not limited? Perhaps we could clarify more on that?

Reply:
Yeah, sorry. Confusing. These are all examples of the high-level algorithm in progress in cases where the original sequence is split twice, resulting in three subsequences, By "first split" I mean, "the state after the first split (due to the first separator)". But my ASCII-art doesn't even make it clear what parts are the returned subsequences, and what part is remaining to be split. I will make this clearer, and also probably remove the whole comment and rewrite it into /Guides/Split.md.